text/0000-decouple-dependencies.md

```diff
@@ -0,0 +1,403 @@
+
+It is recommended that buildpack authors use semver as the type because
+matching is generally simpler that way, however, if a dependency does not
+follow semver you may use regular expressions to match the versions that are
+compatible with the buildpack.
+
+If any semver range or regular expression matches then it can be assumed that
+the buildpack is compatible. If no range matches then it can be assumed that
+the version is not compatible.
+
+For example:
+```toml
+[[metadata.validations]]
+dependency-id = "com.example.jre"
+supported = [ "8.0.*", "11.0.*", "17.0.*" ]
+
+[[metadata.validations]]
+dependency-id = "com.example.nodejs"
+supported = [ "^16.0", "^17.0", "^18.0" ]
+
+[[metadata.validations]]
+dependency-id = "com.example.tomcat"
+supported = [ "8\.5\.\d+", "9\.0\.\d+", "10\.0\.\d+" ]
+type = “regex”
+```
+
+A buildpack is encouraged to be as permissive as possible. This ensures that a
+buildpack author won’t have to frequently update this metadata. This should be
+balanced by buildpack authors to provide compatibility guarantees with the
+tools required to run the buildpack and for the software it is installing to
+run. Generally, we believe that most buildpacks will be compatible with the
+major versions of software they presently support and that new major versions
+of dependencies should be tested and validations should be expanded after
+testing.
+
+## Metadata Distribution
+There is no prescribed method for distributing metadata. It could be done in a
+variety of ways, including HTTPS/SFTP distributed archives, `rsync` of remote
+directories, or even distributed as an image through an image registry.
+
+For the Paketo project, this proposal suggests distributing metadata through
+images in an image registry. This allows the project to use the existing image
+registry to distribute the metadata. Image registries also have inherent
+properties that help with security, like an image cannot be modified without
+creating a new hash for the image and images can be signed (signing is out of
+scope for this proposal). In addition, images are easily versioned such that
+users can hold back updates to dependencies if desired and are easily cached.
+
+The contents of the image will contain the directory structure defined in
+[Metadata Format](#metadata-format). There should not be a top-level directory
+added, so the root of the image should contain all of the directories created
+as top-level organization names, like `com` or `org`. All of the metadata is to
+be included in a single layer. Updates to metadata will require downloading the
+entire layer again, however, it is a single layer and the size is expected to
+be small so this should be very fast.
+
+Further, this proposal suggests subdividing metadata images by project
+sub-team. Each sub-team will be given a unique reverse domain name like
+`io.paketo.java` and `io.paketo.utilities`. In this way, the project’s metadata
+can be easily combined without having conflicts.
```

Comment:
Is this reverse domain name about the metadata image names or the file system structure for dependencies?

Reply:
I read it as filesystem layout - so that multiple metadata sources are guaranteed not to clash on the filesystem. I don't think the image name needs to be specified here, as this RFC doesn't require images to be the distribution mechanism of choice for dependencies.

Reply:
I think that seems reasonable. I guess it comes down to how do you want to ensure that dependencies don't step on each other toes. Language-maintainers was called out in the RFC and seems like an obvious choice. But I guess pre-existing reverse domain name layout (which is common in the java ecosystem) seems reasonable too.

Reply:
Why wouldn't we use the reverse domain names fitting to the upstream dependencies - at least where we don't compile them ourselves, there's nothing Paketo specific to the dependencies, or is there?

Regarding the packaging, if we allow the tree to be represented in a single toml file, grouping similar dependencies should be easy as well.

Reply:
+1, I was also wondering why we would have a paketo-scoped domain for metadata for dependencies that we don't compile

Reply:
The intent is to group by dependency provider. If an upstream project published buildpack dependency metadata, then they could use their own namespace. We publish this metadata, so it goes under `io.paketo...`.

The only place this falls down a bit is when you want to override dependencies for Paketo buildpacks, then you need to republish something under the `paketo.io` namespace, at least initially (see my other comments on how buildpacks can find dependencies). I think that's OK for MVP though, and that there's room here to build and do more sophisticated things so that `com.sap` namespace can publish and override dependencies published by `io.paketo` without having to change the buildpacks.